### PR TITLE
AutoDJ gap fix 

### PR DESCRIPTION
**Pull request by daschuer**
Add a sql querry, that removes deleted and purged tracks from the playlists to get back to a consecutive numbering. 

This is a band aid fixing Bug https://bugs.launchpad.net/mixxx/+bug/1663748 until we apply the crates refactoring here as well.

### CODE REVIEW

File: src/library/basesqltablemodel.h [hunk 1]
```diff
@@ -101,6 +101,8 @@ class BaseSqlTableModel : public QAbstractTableModel, public TrackModel {
     TrackId m_previewDeckTrackId;
     QString m_tableOrderBy;
 
+    QString m_tableName;
```
Comment (rryan) [inline src/library/basesqltablemodel.h]: seems unneeded?

File: src/library/autodj/autodjprocessor.cpp [hunk 2]
```diff
@@ -76,7 +76,7 @@ AutoDJProcessor::AutoDJProcessor(QObject* pParent,
           m_transitionTime(kTransitionPreferenceDefault),
           m_nextTransitionTime(kTransitionPreferenceDefault) {
     m_pAutoDJTableModel = new PlaylistTableModel(this, pTrackCollection,
-                                                 "mixxx.db.model.autodj");
+                                                 "mixxx.db.model.autodj"); //show all tracks
```
Comment (rryan) [inline src/library/autodj/autodjprocessor.cpp]: this comment seems irrelevant to this line of code?

File: src/library/playlisttablemodel.cpp [hunk 3]
```diff
@@ -23,6 +24,15 @@ void PlaylistTableModel::setTableModel(int playlistId) {
     }
 
     m_iPlaylistId = playlistId;
+
+    if (!m_showAll) {
+        // We drop files that have been explicitly deleted from mixxx
```
Comment (rryan) [inline src/library/playlisttablemodel.cpp]: deletd -> deleted

File: src/library/playlisttablemodel.cpp [hunk 4]
```diff
@@ -138,6 +142,31 @@ void PlaylistTableModel::removeTracks(const QModelIndexList& indices) {
     m_pTrackCollection->getPlaylistDAO().removeTracksFromPlaylist(m_iPlaylistId,trackPositions);
 }
 
+void PlaylistTableModel::removeHiddenTracks() {
+    // This query deletes all tracks marked as deletd and all
```
Comment (rryan) [inline src/library/playlisttablemodel.cpp]: deletd -> deleted

**Comment by daschuer** [comment]
fixed!

File: src/library/playlisttablemodel.cpp [hunk 5]
```diff
@@ -36,20 +45,13 @@ void PlaylistTableModel::setTableModel(int playlistId) {
             // the same value as the cover hash.
             << LIBRARYTABLE_COVERART_HASH + " AS " + LIBRARYTABLE_COVERART;
 
-    // We drop files that have been explicitly deleted from mixxx
```
Comment (rryan) [inline src/library/playlisttablemodel.cpp]: Is this comment correct? Do we delete tracks from playlists when we hide them? I thought we only removed tracks from playlists on purge.

**Comment by daschuer** [comment]
> Is this comment correct? Do we delete tracks from playlists when we hide them? I thought we only removed tracks from playlists on purge.

I think the comment was not correct and it is still not fully correct. The issue is that we do not have a good idea to manage hidden tracks in playlists consuming a number, but do not have a handle for the user to manage them.

I propose this solution: 
* Remove hidden and purged tracks from all playlist immediately
* Fix leftover tracks from previous versions (this PR) 
* Keep hidden track in crates. There is no confusion about numbering, so we can look at crates as a kind of tagging system, where the crate membership is a track property, which is kept over a hide unhide cycle.

@uklotzde: does it fit to your view?

**Comment by uklotzde** [comment]
Hidden tracks may cause a lot of headaches and require special case 
handling. For crates only the crate summary properties are affected, which 
we isolated in a separate view. I agree with you that it is not worth to 
keep them in playlist, because otherwise you need to keep track of their 
position somehow. And what is the strategy if you reorder playlists with 
hidden tracks that later become visible again??? We don't have a solution 
that doesn't cause any surprises for the user.

Let's be pragmatic, remove hidden tracks from playlists immediately and 
allow only visible tracks in playlists.

**Comment by daschuer** [comment]
OK, now the playlist gap issue should be solved :-)

**Comment by rryan** [comment]
> Remove hidden and purged tracks from all playlist immediately

I don't think we should delete the track from the playlist until it's actually purged.

**Comment by rryan** [comment]
The problem is in our name for this feature. "hide" isn't the same thing as "remove". Remove implies an irreversible action, hide implies a reversible action. I'd be surprised if hiding caused lost data.

**Comment by daschuer** [comment]
The main use-case for hide, is to remove unwanted tracks like audible books from the library without the library scanner is re-adding it over and over again. An unresolveable gap in the playlist is the bug I am trying to fix. 

For me the loss of playlist membership is the smaller issue. What is the case where this is real problem, I can think of an undo loop ...

@rryan: any idea how to improve the situation?

**Comment by rryan** [comment]
> @rryan: any idea how to improve the situation?

What about warning the user on hide that the track will be removed from playlists / crates?

**Comment by daschuer** [comment]
Yes, normally I am against pop-up boxes, but this will only appear if the user has a track added to a playlist.  The use-case is when you want to hide a mp3 track which has also a flac version.

**Comment by uklotzde** [comment]
Tracks are hidden, because they are supposed to not appear in the library at all and not only in the tracks section. It would even be sufficient if Mixxx only keeps track of their file names, but keeping them like regular tracks in the library is simpler for displaying them.

Asking for confirmation to delete the references of to be hidden tracks from crates and playlists sounds good. This will avoid a lot of confusion caused by this half-baked feature in the past.

How do we handle the required cleanup task? We should think about increasing the database version even if the schema itself does not change. Cleanup for crates is trivial, but I'm not sure if renumbering the playlist items can solely be accomplished with some SQL commands.

**Comment by daschuer** [comment]
Now the requested dialog is in place. The gap in the playlist remains, 

The gap issue still remain, but unlike before the user is able to fill the gap. 
Still not perfect, but better than an unfulfillable gap like we had before.  

This is an issue that should be solved as well, but probably in a separate PR?

File: src/library/dao/playlistdao.cpp [hunk 6]
```diff
@@ -380,15 +380,60 @@ bool PlaylistDAO::isHidden(const int playlistId) const {
...
+    }
+    transaction.commit();
+    emit(changed(playlistId));
+}
 
+void PlaylistDAO::removeTracksFromPlaylistInner(int playlistId, int position) {
+    QSqlQuery query(m_database);
     query.prepare("SELECT track_id FROM PlaylistTracks WHERE playlist_id=:id "
-                  "AND position=:position");
+                "AND position=:position");
```
Comment (rryan) [inline src/library/dao/playlistdao.cpp]: BTW, this is the correct formatting for a hanging indent per the Google C++ Style Guide. What you've changed it to is only correct in this case:

```c++
query.prepare(
        "....");
```

File: src/library/trackcollection.cpp [hunk 7]
```diff
@@ -143,6 +143,38 @@ void TrackCollection::relocateDirectory(QString oldDir, QString newDir) {
 }
 
 bool TrackCollection::hideTracks(const QList<TrackId>& trackIds) {
+    // Warn if tracks have a playlist membership
+    QSet<int> allPlaylisIds;
```
Comment (rryan) [inline src/library/trackcollection.cpp]: playlis -> playlist (and below)

File: src/library/trackcollection.cpp [hunk 8]
```diff
@@ -143,6 +143,38 @@ void TrackCollection::relocateDirectory(QString oldDir, QString newDir) {
...
+    if (allPlaylisIds.count()) {
+        QString playlistNames;
+        for (const auto& playlisId: allPlaylisIds) {
+            playlistNames += "\"" + m_playlistDao.getPlaylistName(playlisId) + "\"\n";
+        }
+
+        if (QMessageBox::question(
+                nullptr,
+                tr("Hiding tracks"),
+                tr("The selected tracks are member of the following playlists:\n\n"
```
Comment (rryan) [inline src/library/trackcollection.cpp]: A little clearer:

The selected tracks are in the following playlists: %1. 
Hiding them from the library will remove them from these playlists. Continue?

It would also be nice to have the newlines out of hte translation string. Maybe they could be part of the %1.

**Comment by daschuer** [comment]
Done. Now the playlist is also renumbered when there was a gap issue from a previous version.

File: src/library/dao/playlistdao.cpp [hunk 9]
```diff
@@ -380,13 +380,96 @@ bool PlaylistDAO::isHidden(const int playlistId) const {
...
 
-void PlaylistDAO::removeTrackFromPlaylist(const int playlistId, const int position)
-{
+
+void PlaylistDAO::removeHiddenTracks(const int playlistId) {
+	ScopedTransaction transaction(m_database);
+    // This query deletes all tracks marked as deleted and all
+    // phantom track_ids with no match in the library table
+    QString queryString = QString(
+            "SELECT position FROM PlaylistTracks "
```
Comment (uklotzde) [inline src/library/dao/playlistdao.cpp]: Why don't you execute "DELETE FROM PlaylistTracks..." directly? Afterwards you can adjust the position of all remaining tracks in this playlist. Or am I missing something here?

Adjusting the position of playlist tracks should be common operation. You may also provide a start position (which is known to be valid, default = 0) and adjust only the positions greater than this start position.

Comment (daschuer) [inline src/library/dao/playlistdao.cpp]: Doing it like this was the easiest way for me because I could reuse existing code, without much testing. My original idea was just to fix the issue of AutoDJ and postpone the rest ... now it is even better but not perfect. 

I have not thought much about a good way to close all gaps and find duplicates in an effective way. 
I can only think of a query, sort by number and compare number with index. If It does not match, fix it. Such a code would be required to fix issues in old libraries. 

Is it OK to postpone this to an other PR?

**Comment by uklotzde** [comment]
I didn't notice that you are reusing existing functionality. Let's postpone 
the refactoring to keep the change set limited, one step at a time.

File: src/library/trackcollection.cpp [hunk 10]
```diff
@@ -143,6 +143,38 @@ void TrackCollection::relocateDirectory(QString oldDir, QString newDir) {
 }
 
 bool TrackCollection::hideTracks(const QList<TrackId>& trackIds) {
+    // Warn if tracks have a playlist membership
+    QSet<int> allPlaylistIds;
+    for (const auto& trackId: trackIds) {
+        QSet<int> playlisIds;
```
Comment (uklotzde) [inline src/library/trackcollection.cpp]: typo

File: src/library/trackcollection.cpp [hunk 11]
```diff
@@ -143,6 +143,38 @@ void TrackCollection::relocateDirectory(QString oldDir, QString newDir) {
...
+        m_playlistDao.getPlaylistsTrackIsIn(trackId, &playlisIds);
+        QString playlistNames;
```
Comment (uklotzde) [inline src/library/trackcollection.cpp]: unused

File: src/library/trackcollection.cpp [hunk 12]
```diff
@@ -143,6 +143,38 @@ void TrackCollection::relocateDirectory(QString oldDir, QString newDir) {
...
+        for (const auto& playlisId: playlisIds) {
+            if (m_playlistDao.getHiddenType(playlisId) != PlaylistDAO::PLHT_SET_LOG) {
+                allPlaylistIds.insert(playlisId);
+            }
+        }
+    }
+
+    if (allPlaylistIds.count()) {
```
Comment (uklotzde) [inline src/library/trackcollection.cpp]: !isEmpty() is more expressive

File: src/library/trackcollection.cpp [hunk 13]
```diff
@@ -143,6 +143,38 @@ void TrackCollection::relocateDirectory(QString oldDir, QString newDir) {
...
+        QString playlistNames = "\n\n";
```
Comment (uklotzde) [inline src/library/trackcollection.cpp]: I would collect the names of all playlists in a QStringList that is formatted when building the string for the message box text. Slightly less efficient, but I prefer clean and expressive code.

File: src/library/dao/playlistdao.cpp [hunk 14]
```diff
@@ -723,54 +751,11 @@ int PlaylistDAO::getMaxPosition(const int playlistId) const {
...
-    if (!query.exec()) {
-        LOG_FAILED_QUERY(query);
-        return;
-    }
-
-    foreach (int playlistId, removedTracksPlaylistIds) {
-        emit(changed(playlistId));
+    	const auto it = m_playlistsTrackIsIn.find(trackId);
+    	while (it != m_playlistsTrackIsIn.end() && it.key() == trackId) {
+    		removeTrackFromPlaylist(it.value(), trackId);
```
Comment (uklotzde) [inline src/library/dao/playlistdao.cpp]: The code does not reveal why it is safe to keep iterating while m_playlistsTrackIsIn is modified in removeTrackFromPlaylist(). Please check that this is true and add a comment. Or take the safe route and store the result of the iteration in a temporary collection. I would recommend the latter.

File: src/library/crate/cratetablemodel.cpp [hunk 15]
```diff
@@ -32,10 +32,10 @@ void CrateTableModel::selectCrate(CrateId crateId) {
             // For sorting the cover art column we give LIBRARYTABLE_COVERART
             // the same value as the cover hash.
             << LIBRARYTABLE_COVERART_HASH + " AS " + LIBRARYTABLE_COVERART;
-    // We drop files that have been explicitly deleted from mixxx
-    // (mixxx_deleted=0) from the view. There was a bug in <= 1.9.0 where
-    // removed files were not removed from crates, so some users will have
-    // libraries where this is the case.
+    // We hide files that have been explicitly deleted in the library
+    // (mixxx_deleted = 0) from the view.
+    // They are kept in the database, because we treat crate memershipas a
```
Comment (uklotzde) [inline src/library/crate/cratetablemodel.cpp]: typo

File: src/library/crate/cratestorage.cpp [hunk 16]
```diff
@@ -174,25 +178,29 @@ void CrateStorage::repairDatabase(QSqlDatabase database) {
...
         FwdSqlQuery query(database, QString(
                 "DELETE FROM %1 WHERE %2 NOT IN (SELECT %3 FROM %4)").arg(
                         CRATE_TRACKS_TABLE,
                         CRATETRACKSTABLE_TRACKID,
                         LIBRARYTABLE_ID,
                         LIBRARY_TABLE));
         if (query.execPrepared() && (query.numRowsAffected() > 0)) {
-            qWarning() << "Deleted" << query.numRowsAffected() << "crate tracks of non-existent tracks";
+            qWarning() << "Removed" << query.numRowsAffected()
+                    << "library purged tracks form crates";
```
Comment (uklotzde) [inline src/library/crate/cratestorage.cpp]: typo

File: src/library/crate/cratestorage.cpp [hunk 17]
```diff
@@ -174,25 +178,29 @@ void CrateStorage::repairDatabase(QSqlDatabase database) {
 
     // Crate tracks
     {
+        // Remove tracks form non-existent crates
```
Comment (uklotzde) [inline src/library/crate/cratestorage.cpp]: Multiple typos "form" -> "from"

File: src/library/crate/cratestorage.cpp [hunk 18]
```diff
@@ -139,15 +139,18 @@ void CrateStorage::repairDatabase(QSqlDatabase database) {
...
                         CRATE_TABLE,
                         CRATETABLE_NAME));
         if (query.execPrepared() && (query.numRowsAffected() > 0)) {
-            qWarning() << "Deleted" << query.numRowsAffected() << "crates with empty names";
+            qWarning() << "Deleted" << query.numRowsAffected()
+                    << "crates with empty names";
         }
     }
     {
+        // Fix invalid values in the "locked" column;
```
Comment (uklotzde) [inline src/library/crate/cratestorage.cpp]: Remove the finishing semicolon from all comments ;)

**Comment by daschuer** [comment]
notes addressed

**Comment by daschuer** [comment]
osx build fails because of time out 
```
The job exceeded the maximum time limit for jobs, and has been terminated.
```

**Comment by daschuer** [comment]
Can this be merged now?

File: src/library/trackcollection.cpp [hunk 19]
```diff
@@ -143,6 +143,42 @@ void TrackCollection::relocateDirectory(QString oldDir, QString newDir) {
 }
 
 bool TrackCollection::hideTracks(const QList<TrackId>& trackIds) {
+    // Warn if tracks have a playlist membership
+    QSet<int> allPlaylistIds;
+    for (const auto& trackId: trackIds) {
+        QSet<int> playlistIds;
+        m_playlistDao.getPlaylistsTrackIsIn(trackId, &playlistIds);
+        for (const auto& playlisId: playlistIds) {
```
Comment (uklotzde) [inline src/library/trackcollection.cpp]: typo

**Comment by uklotzde** [comment]
Just one more typo ;)

**Comment by daschuer** [comment]
Thank you for review.

**Comment by uklotzde** [comment]
@daschuer: There seems to be a regression caused by this changes
https://bugs.launchpad.net/mixxx/+bug/1671247

**Comment by daschuer** [comment]
Thank you for the hint. 
It is fixed now in https://github.com/mixxxdj/mixxx/commit/3eb8844a5e5a789395cd0d72b45040c38521697d
It seams to have worked magically due to the bug of having no local copy of the multi hash.
